File: boj/PrefixSum/BOJ_16507.py
# 백준 16507 어두운 건  무서워
# 22.12.08
# https://github.com/skyqnaqna/algorithm_study

import sys
rd = sys.stdin.readline

r, c, q = map(int, rd().split())

prefixSum = [[0] * (c + 1) for _ in range(r + 1)]

for i in range(1, r + 1):
    brightness = list(map(int, rd().split()))

    for j in range(1, c + 1):
        prefixSum[i][j] += prefixSum[i][j - 1] + brightness[j - 1]

for j in range(1, c + 1):
    for i in range(2, r + 1):
        prefixSum[i][j] += prefixSum[i - 1][j]


for _ in range(q):
    r1, c1, r2, c2 = map(int, rd().split())

    cnt = (r2 - r1 + 1) * (c2 - c1 + 1)

    res = (prefixSum[r2][c2] - prefixSum[r1 - 1][c2] - prefixSum[r2][c1 - 1] + prefixSum[r1 - 1][c1 - 1]) // cnt

    print(res)

File: boj/PrefixSum/BOJ_16507.swift
/*
 백준 16507 어두운 건 무서워
 22.12.08
 https://github.com/skyqnaqna/algorithm_study
*/

import Foundation

let input = readLine()!.split(separator: " ").map { Int(String($0))! }
let r = input[0], c = input[1], q = input[2]

var prefixSum = [[Int]](repeating: [Int](repeating: 0, count: c + 1), count: r + 1)

for i in 1 ... r {
  let brightness = readLine()!.split(separator: " ").map { Int(String($0))! }

  for j in 1 ... c {
    prefixSum[i][j] += prefixSum[i][j - 1] + brightness[j - 1]
  }
}

for j in 1 ... c {
  for i in 1 ... r {
    prefixSum[i][j] += prefixSum[i - 1][j]
  }
}

var answer = ""

for _ in 0 ..< q {
  let i = readLine()!.split(separator: " ").map { Int(String($0))! }

  let r1 = i[0], c1 = i[1], r2 = i[2], c2 = i[3]

  let cnt = (r2 - r1 + 1) * (c2 - c1 + 1)

  let res = (prefixSum[r2][c2] - prefixSum[r1 - 1][c2] - prefixSum[r2][c1 - 1] + prefixSum[r1 - 1][c1 - 1]) / cnt

  answer += "\(res)\n"
}

print(answer)
